File: leetcode/leetcode/020.cpp
//
//  020.cpp
//  leetcode
//
//  Created by Hang Zhang on 4/28/15.
//  Copyright (c) 2015 Hilings Studio. All rights reserved.
//

#include <iostream>
#include <string>
#include <stack>
using namespace std;

class Solution {
public:
    bool isValid(string s) {
        stack<char> stack;
        for (int i = 0; i < s.size(); i++) {
            if (stack.empty())
                stack.push(s[i]);
            else if ( (s[i] == ')' && stack.top() == '(') || (s[i] == ']' && stack.top() == '[') || (s[i] == '}' && stack.top() == '{') )
                stack.pop();
            else
                stack.push(s[i]);
        }
        return stack.empty();
    }
};

int main(int arg, char *argv[]) {
    // insert code here...
    cout << "LeetCode 020 Valid Parentheses, C++ ... ...\n";
    
    string s = "]";
    cout << s << '\n';
    
    Solution sol;
    cout << (sol.isValid(s) ? "True" : "False") << '\n';
    
    return 0;
}

File: 020.py
#   020
#   Valid Parentheses
#   2015-04-28
#####################################################
class Solution:
    # @param {string} s
    # @return {boolean}
    def isValid(self, s):
        stack = []
        for c in s:
            if not stack:
                stack.append(c)
            elif (c == ')' and stack[-1] == '(') or (c == ']' and stack[-1] == '[') or (c == '}' and stack[-1] == '{'):
                stack.pop()
            else:
                stack.append(c)
        return not stack

#####################################################

s = "]"
print s
sol = Solution()
print "True" if sol.isValid(s) else "False"

